net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+	struct m0_net_buffer   *nb = fb->fb_nb;
+
+	fb->fb_rbd = (struct m0_fab__bdesc *)(nb->nb_desc.nbd_data);
+	fb->fb_riov = (struct fi_rma_iov *)(nb->nb_desc.nbd_data + 
+					    sizeof(struct m0_fab__bdesc));
+	libfab_ep_ntop(fb->fb_rbd->fbd_netaddr, epname);
+	M0_ASSERT(fb->fb_rbd->fbd_iov_cnt <= FAB_IOV_MAX);
+}
+
+/**
+ * Register the buffer with the appropriate access to the domain
+ */
+static int libfab_buf_dom_reg(struct m0_net_buffer *nb, struct fid_domain *dp)
+{
+	struct m0_fab__buf    *fbp = nb->nb_xprt_private;
+	struct m0_fab__buf_mr *mr = &fbp->fb_mr;
+	uint64_t               key;
+	int                    seg_nr = nb->nb_buffer.ov_vec.v_nr;
+	int                    i;
+	int                    ret = FI_SUCCESS;
+	uint32_t               retry_cnt;
+
+	M0_PRE(fbp != NULL && dp != NULL);
+	M0_PRE(seg_nr <= FAB_IOV_MAX);
+
+	if (fbp->fb_dp == dp)
+		return M0_RC(ret);
+
+	for (i = 0; i < seg_nr; i++) {
+		retry_cnt = 20;
+		key = libfab_mr_keygen();
+		
+		ret = fi_mr_reg(dp, nb->nb_buffer.ov_buf[i], 
+				nb->nb_buffer.ov_vec.v_count[i],
+				FAB_MR_ACCESS, FAB_MR_OFFSET, key,
+				FAB_MR_FLAG, &mr->bm_mr[i], NULL);
+
+		if (ret != FI_SUCCESS) {
+			M0_LOG(M0_ERROR, "fi_mr_reg = %d i =%d key = %"PRIu64,
+			       ret, i, key);
+			while(ret != FI_SUCCESS && retry_cnt > 0)
+			{
+				key = libfab_mr_keygen();
+				ret = fi_mr_reg(dp, nb->nb_buffer.ov_buf[i],
+						nb->nb_buffer.ov_vec.v_count[i],
+						FAB_MR_ACCESS, FAB_MR_OFFSET,
+						key, FAB_MR_FLAG, &mr->bm_mr[i],
+						NULL);
+				retry_cnt--;
+			}
+			M0_ASSERT(ret == FI_SUCCESS);
+		}
+
+		mr->bm_desc[i] = fi_mr_desc(mr->bm_mr[i]);
+	}
+
+	fbp->fb_dp = dp;
+
+	return M0_RC(ret);
+}
+
+/**
+ * Get the destination address from the endpoint name
+ */
+static inline uint64_t libfab_destaddr_get(struct m0_fab__ep_name *epname)
+{
+	uint64_t dst;
+
+	libfab_ep_pton(epname, &dst);
+	dst |= 0x02;	/* FI_SOCKADDR_IN */
+	return dst;
+}
+
+/**
+ * Buffers are queued before connection establishment, send those buffers
+ * after connection establishment
+ */
+static void libfab_pending_bufs_send(struct m0_fab__ep *ep)
+{
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__buf       *fbp;
+	struct m0_net_buffer     *nb;
+	struct iovec              iv;
+	int                       ret = 0;
+
+	aep = libfab_aep_get(ep);
+	m0_tl_for(fab_sndbuf, &ep->fep_sndbuf, fbp) {
+		nb = fbp->fb_nb;
+		fbp->fb_txctx = ep;
+		iv.iov_base = nb->nb_buffer.ov_buf[0];
+		iv.iov_len = nb->nb_buffer.ov_vec.v_count[0];
+		switch (nb->nb_qtype) {
+			case M0_NET_QT_MSG_SEND:
+				ret = fi_sendv(aep->aep_txep, &iv,
+					       fbp->fb_mr.bm_desc, 1, 0, fbp);
+				break;
+			case M0_NET_QT_ACTIVE_BULK_RECV:
+			case M0_NET_QT_ACTIVE_BULK_SEND:
+				fbp->fb_wr_cnt = libfab_wr_cnt_get(fbp);
+				libfab_bulklist_add(libfab_buf_ma(nb), fbp,
+						    aep);
+				break;
+			default: M0_LOG(M0_ERROR, "Invalid qtype=%d",
+					nb->nb_qtype);
+				break;
+		}
+		fab_sndbuf_tlist_del(fbp);
+		M0_ASSERT(ret == FI_SUCCESS);
```

Comment:
Should we handle error here instead of assert?

Reply:
Error handling added